packages/nitro-protocol/contracts/MultiAssetHolder.sol

```diff
@@ -430,7 +430,7 @@ contract MultiAssetHolder is IMultiAssetHolder, StatusManager {
                         // reduce the new allocationItem.amount
                         newAllocation[i].amount -= affordsForDestination;
                         // increase the relevant payout
-                        payouts[k] += affordsForDestination;
+                        payouts[i] += affordsForDestination;
```

Comment:
I've been looking at this for a good few minutes and maybe I've confused myself, so I'll put the question forward rather than keep trying to see through to the answer.

Given that `payouts` is now the same length as the allocations array, can't some of this logic (ie, 428 check, k-indexing) be smoothed out?

Reply:
I think there's some unavoidable complexity here in having to track an index for i) the guarantee, ii) the target and iii) the `indices` themselves.

There was some avoidable complexity in having the payouts be two different lengths based on `indices`. We've removed that here. 

That's not to say we can't keep improving this!

Reply:
@geoknee I'm not sure I understand your answer here.

After making `payouts.length` equal to `allocation.length`, we are:
1. iterating through each allocation item at index `i`
2. making sure that `k == i`
3. setting `payouts[k]` depending on whether something should be payed out to `allocations[i].destination`

I believe Colin is suggesting that we remove the point where we ensure `k == i`, resulting in this diff:
```
diff --git a/packages/nitro-protocol/contracts/MultiAssetHolder.sol b/packages/nitro-protocol/contracts/MultiAssetHolder.sol
index 9cf170d51..4a35c43fb 100644
--- a/packages/nitro-protocol/contracts/MultiAssetHolder.sol
+++ b/packages/nitro-protocol/contracts/MultiAssetHolder.sol
@@ -402,7 +402,6 @@ contract MultiAssetHolder is IMultiAssetHolder, StatusManager {
         totalPayouts = 0;
         allocatesOnlyZeros = true; // switched to false if there is an item remaining with amount > 0
         uint256 surplus = initialHoldings; // tracks funds available during calculation
-        uint256 k = 0; // indexes the `indices` array
         //  We rely on the assumption that the indices are strictly increasing.
         //  This allows us to iterate over the destinations in order once, continuing until we hit the first index, then the second etc.
         //  If the indices were to decrease, we would have to start from the beginning: doing a full search for each index.
@@ -425,16 +424,14 @@ contract MultiAssetHolder is IMultiAssetHolder, StatusManager {
                     uint256 affordsForDestination = min(allocation[i].amount, surplus);
                     // decrease surplus by the current amount regardless of hitting a specified index
                     surplus -= affordsForDestination;
-                    if ((indices.length == 0) || ((k < indices.length) && (indices[k] == i))) {
-                        // only if specified in supplied indices, or we if we are doing "all"
-                        // reduce the new allocationItem.amount
-                        newAllocation[i].amount -= affordsForDestination;
-                        // increase the relevant payout
-                        payouts[i] += affordsForDestination;
-                        totalPayouts += affordsForDestination;
-                        // move on to the next supplied index
-                        ++k;
-                    }
+                    // only if specified in supplied indices, or we if we are doing "all"
+                    // reduce the new allocationItem.amount
+                    newAllocation[i].amount -= affordsForDestination;
+                    // increase the relevant payout
+                    payouts[i] += affordsForDestination;
+                    totalPayouts += affordsForDestination;
+                    // move on to the next supplied index
+
                     break; // start again with the next guarantee destination
                 }
             }
```

Reply:
https://github.com/statechannels/statechannels/pull/3742

Reply:
ok, so I understand a bit more now:
`(indices.length == 0) || ((k < indices.length) && (indices[k] == i))` does not always evaluate to `true`, as I previously believed it to.

Instead, it evaluates to `true` when the index `i` is "the next" index supplied by the user to execute.

Reply:
@NiloCK do you think [this](https://github.com/statechannels/statechannels/pull/3743) clarifies things?